tee-worker/litentry/core/assertion-build/src/a4.rs

```diff
@@ -102,98 +102,103 @@ pub fn build(
-				|| verified_network == VerifiedCredentialsNetwork::Litmus
-				|| verified_network == VerifiedCredentialsNetwork::LitentryRococo
-			{
-				(min_balance / (10 ^ 12)) as f64
-			} else {
-				(min_balance / (10 ^ 18)) as f64
-			};
-
-			let mut addresses: Vec<String> = vec![];
-			match &identity {
-				Identity::Evm { address, .. } => {
-					let mut address = account_id_to_string(address.as_ref());
-					address.insert_str(0, "0x");
-					debug!("	[AssertionBuild] A4 EVM address : {}", address);
-
-					addresses.push(address);
-				},
-				Identity::Substrate { address, .. } => {
-					let mut address = account_id_to_string(address.as_ref());
-					address.insert_str(0, "0x");
-					debug!("	[AssertionBuild] A4 Substrate address : {}", address);
+				if_match_networks_collect_address(&mut networks, (*network).into(), address);
+			},
+			Identity::Evm { network, address } => {
+				let mut address = account_id_to_string(address.as_ref());
+				address.insert_str(0, "0x");
 
-					addresses.push(address);
-				},
-				Identity::Web2 { address, .. } => match from_utf8(address.as_ref()) {
-					Ok(addr) => addresses.push(addr.to_string()),
-					Err(e) => error!(
-						"	[AssertionBuild] A4 parse error Web2 address {:?}, {:?}",
-						address, e
-					),
+				if_match_networks_collect_address(&mut networks, (*network).into(), address);
+			},
+			_ => {},
+		};
+	});
+
+	let mut is_hold = false;
+	let mut optimal_hold_index = ASSERTION_FROM_DATE.len() - 1;
+
+	// If both Substrate and Evm networks meet the conditions, take the interval with the longest holding time.
+	// Here's an example:
+	//
+	// ALICE holds 100 LITs since 2018-03-02 on substrate network
+	// ALICE holds 100 LITs since 2020-03-02 on evm network
+	//
+	// min_amount is 1 LIT
+	//
+	// the result should be
+	// Alice:
+	// [
+	//    from_date: < 2019-01-01
+	//    to_date: >= 2023-03-30 (now)
+	//    value: true
+	// ]
+	for (verified_network, addresses) in networks {
+		let addresses: Vec<String> = addresses.into_iter().collect();
+		let token_address = if verified_network == VerifiedCredentialsNetwork::Ethereum {
+			LIT_TOKEN_ADDRESS
+		} else {
+			""
+		};
+
+		// TODO:
+		// There is a problem here, because TDF does not support mixed network types,
+		// It is need to request TDF 2 (substrate+evm networks) * 7 (ASSERTION_FROM_DATE) = 14 http requests.
+		// If TDF can handle mixed network type, and even supports from_date array,
+		// so that ideally, up to one http request can yield results.
+		for (index, from_date) in ASSERTION_FROM_DATE.iter().enumerate() {
+			let vch = VerifiedCredentialsIsHodlerIn::new(
+				addresses.clone(),
+				from_date.to_string(),
+				verified_network.clone(),
+				token_address.to_string(),
+				q_min_balance.to_string(),
+			);
+			match client.check_verified_credentials_is_hodler(vch) {
+				Ok(is_hodler_out) => {
+					for hodler in is_hodler_out.verified_credentials_is_hodler.iter() {
+						is_hold = is_hold || hodler.is_hodler;
```

Comment:
I think you need to reset `is_hold` in `for (verified_network, addresses) in networks` - otherwise the second network would always start with `is_hold == true`

Reply:
Yes, should reset is_hold to false every networks loop.